### ISSUE TITLE
Using Recorder to record custom traces for agent Pre-Training

### ISSUE THREAD

**Issue by IbraheemNofal** [tensorforce/tensorforce#708]
As discussed in the [Gitter thread](https://gitter.im/tensorforce/community), I'm interested in the pre-training feature. Now it's been made clear to me that it's normally used to cut down on re-training time when I've already got a trained agent, in which case I could use the recorder feature when instantiating the already trained agent to record traces, and then use `agent.pretrain()` which does the following:
for `num_iterations` times, it loads `num_traces` trace files (which each contain `recorder['frequency']` episodes), feeds them via `agent.experience()` to the agent's internal memory, and then triggers `num_updates` updates, which use experience sampled from the memory, so what was fed in this (and potentially previous) iterations.

But for my use-case, I don't have a trained agent, but I've got a 'benchmark algorithm' that I want to record training traces for, and then use those traces for pre-training my agent, sort of like 'supervised' reinforcement learning. Does this sound like a valid use-case?
Now, as you've explained, this can be done by recording traces the same way recorder does, by utilizing [`numpy.savez_compressed(**experiences)`](https://numpy.org/doc/stable/reference/generated/numpy.savez_compressed.html), but it isn't clear to me how to structure those 'experiences' when saving. 

1) How exactly are the 'experiences' structured. Is the saving simply done by calling `numpy.savez_compressed('trace-1', states= states_array, actions = actions_array, terminal = terminals_array, reward = rewards_array)`
2) Oh, and as far as structuring those arrays, I'm guessing that for instance at index 0 for each array it means that:
When the agent observes `state = state_array[0]` and takes `action = action_array[0]` it observed `reward = reward_array[0]` and `terminal = terminal[0]`. Correct?

**Comment by AlexKuhnle** [comment]
Both correct, that should work. Moreover, note that the files currently have to be prefixed by `trace-` (will be changed with the next version to suffix `.npz`), and any file order or so is currently ignored (there will be an option to go through experience in sorted order in the future, if that should be preferable). Finally, regarding the content, trace files have to consist of (potentially multiple) full episodes (terminated by `terminal=True/1/2`) which are simply concatenated. It is important that the timesteps are not just independent demonstration steps (like a supervised learning dataset, independent instances), but together form an "episode".

Hope that helps! Looking forward to hear how it goes...

**Comment by AlexKuhnle** [comment]
Fyi, related issue: https://github.com/tensorforce/tensorforce/issues/700

**Comment by IbraheemNofal** [comment]
I've managed to record my own custom traces and got pre-training working. I have a question though, how exactly does pre-training affect exploration? Does the exploration decay based on the amount of updates done during pre-training? Or is it not affected at all?

**Comment by AlexKuhnle** [comment]
The only counter which is increased for `agent.pretrain()` is the `updates` counter, so if your exploration decay is using `timesteps` or `episodes`, it won't be affected. `Timesteps` and `episodes` only count the regular act-observe interactions.

**Comment by AlexKuhnle** [comment]
Related to the request in #700, it is now possible to record interaction traces for a (non-Agent) act-function, see [docs](https://tensorforce.readthedocs.io/en/latest/basics/features.html#record-pretrain). I think this can be closed now.
